### PR TITLE
Limit num of bytes read in _consume_request.

### LINKED ISSUE #11: _consume_request implementation is vulnerable to denial of service attacks
The current `_consume_request` implementation reads the `Content-Length` header and blindly tries to `read` and buffer that many bytes into memory without regard for size:
https://github.com/mkomitee/wsgi-kerberos/blob/c2199a3c1143c0a340b576c0d547442224080e53/wsgi_kerberos.py#L57

I think it would be safer to only read up to `read_max` bytes.

### LINKED ISSUE #12: _consume_request does not handle Transfer-Encoding: chunked
The current implementation of `_consume_request` assumes a 0-length body when a `Content-Length` header is not sent:
https://github.com/mkomitee/wsgi-kerberos/blob/c2199a3c1143c0a340b576c0d547442224080e53/wsgi_kerberos.py#L49

However, clients may stream request bodies using the `Transfer-Encoding: chunked` without sending a `Content-Length` header.

To handle this, we would just read chunks of data from the socket until hitting`\r\n\r\n`.

Maybe we're fine not handling this, and just giving clients "Broken Pipe" for such requests? If that's the case, the code change here could just be to update the comments around this code to document that decision.

### PR DESCRIPTION
**Pull request by twosigmajab**
Limit num of bytes read in _consume_request.

Fixes #11.

Also, only consume the request for naive clients that do not send the
Expect: 100-continue header.

Finally, when no Content-Length header is sent (e.g. for chunked
Transfer-Encoding), still try to consume the request (up to the limit).
Fixes #12.

### CODE REVIEW

File: wsgi_kerberos.py
Comment (twosigmajab) [inline wsgi_kerberos.py]: Looks like golang uses a read max of 2 KiB (weirdly written as `2 << 10`):
https://github.com/golang/go/blob/ee3dded36d69264998c39af0ec851371850d842b/src/net/http/client.go#L616-L625
Can reduce this default to match that in my next round of addressing review feedback.

Comment (twosigmajab) [reply wsgi_kerberos.py]: Done in the latest revision.

**Review by propertone — CHANGES_REQUESTED** [review]
This change appears to break unit tests for me (test_forbidden hangs in the receive loop)

**Comment by twosigmajab** [comment]
Sorry about that, fixed! Another look?

**Review by propertone — COMMENTED** [review]
Since you're limiting the max length of the data being read, do we still need to read the data in chunks? Presumably, you can read up to 2k in one call.

**Comment by jab** [comment]
@propertone, the behavior of `sock.read(N)` is to "receive **up to** N bytes from the socket", so you can't necessarily read all 2k in one call. (The reason `read` returns the actual number of bytes read is that sometimes it is not able to read all N bytes you requested (even when N is small). So that's how you're able to tell when additional `read` calls are required.)

We could still get rid of the read loop if you prefer to make even less of an effort to prevent affected clients from seeing "broken pipe". Happy to do that if you prefer, just let me know.

**Review by propertone — APPROVED** [review]
I think this is fine as is - thanks for the contribution! Will publish a new release after merging this PR.

**Review by propertone — CHANGES_REQUESTED** [review]
Actually, I just realized one more thing - the golang comparison is faulty because that's the client code, not the server-code.
The server-side code actually has a [10MB limit](https://github.com/golang/go/blob/master/src/net/http/request.go#L1204), which seems more reasonable. Maybe the limit can be a top-level constant, so someone could technically change it if necessary (such case would probably be rarely enough that a class variable is probably unnecessary).

**Comment by twosigmajab** [comment]
🤦 good catch, thanks @propertone! Fixed the default value in the latest revision. Rather than exposing a mutable, global module variable for the case where a user needs to override the default value, I think exposing an optional constructor param is a safer API, even if it is rarely needed.

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: Is that still the right thing to do? I think other web servers just read all data up to max length. On the other hand, if the client will get a connection reset anyway, might as well do it early.

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: The golang ref is TMI - put it as a code comment instead. Also add `:type` for consistency.

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: Trying to think of a slightly more friendly name here...`limit_request_size` ([apache](https://httpd.apache.org/docs/2.4/mod/core.html#limitrequestbody)), `max_body_size` ([nginx](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size)), `max_content_length` ([flask](https://flask.palletsprojects.com/en/1.1.x/config/#MAX_CONTENT_LENGTH))...something more along those lines would be better.

**Review by propertone — CHANGES_REQUESTED** [review]
Few other comments:
* Given the various issues uncovered so far, I'm not confident that this code is correct for various input permutations of body size and content-length. Can you add at least 1 unit test for it?
* One of the issues this PR supposedly addresses is chunked transfer encoding, but have you confirmed that this actually works as intended in practice? I wasn't able to reproduce this since wsgi just reads the entire body.

**Comment by twosigmajab** [comment]
@propertone, thinking about this some more, I actually propose a different approach; please see #25. I don't think we should do this in WSGI-Kerberos, since it is already done in production-quality WSGI servers, and doing it in WSGI-Kerberos defeats that. WSGI servers (or even reverse proxies like NGINX) are the proper place to do this; they can actually implement this more safely, since they have raw socket access.

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: > On the other hand, if the client will get a connection reset anyway, might as well do it early.

Exactly. I left a comment to make this clearer.

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: Done in the latest revision.

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: What do you think of `read_max_on_401`?

I did think carefully about the name for this before I chose it. I think the issue with the three shorter examples you gave is that they don't mean the same thing: they mean an unconditional limit on all requests, not just unauthenticated ones as in this case. I hope `read_max_on_401` is a good compromise between conciseness and preciseness.

File: test_wsgi_kerberos.py
Comment (twosigmajab) [inline test_wsgi_kerberos.py:139]: Added a passing test for this.

File: wsgi_kerberos.py
Comment (twosigmajab) [inline wsgi_kerberos.py:51]: I noticed and fixed this additional issue with the current implementation. Just a few lines above, the quote from the spec says this should only be done for a request that does not include an "Expect: 100-continue" header. We should not attempt to consume *Content-Length* bytes of request bodies sent by clients that send "Expect: 100-continue", since they wait for a "100 Continue" response before actually sending the body.

File: wsgi_kerberos.py
Comment (twosigmajab) [inline wsgi_kerberos.py:60]: (KeyError looks like it was mistakenly left in here after e.g. `environ['CONTENT_LENGTH']` was changed to `environ.get('CONTENT_LENGTH')` above in an earlier revision? And ValueError looks like it's trying to account for the `int(content_length_string)` conversion, but I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here?)

It looks like this should just catch Exception in any case -- this is a best effort function that should not ever cause an unhandled exception and resulting 500 error on the server. I made that improvement below, and also added a debug log statement with the exception info so the user actually has a way of finding out more about any swallowed exceptions.

**Review by twosigmajab — COMMENTED** [review]
I believe the latest revision addresses all the latest review comments. Another look?

**Comment by twosigmajab** [comment]
> One of the issues this PR supposedly addresses is chunked transfer encoding, but have you confirmed that this actually works as intended in practice? I wasn't able to reproduce this since wsgi just reads the entire body.

What servers did you test with? Support was added for this to several popular implementations in the past couple years. If you try with e.g. Gunicorn and Flask I think you should be able to reproduce. Please see e.g. https://github.com/benoitc/gunicorn/issues/1653 (which I also linked from https://github.com/deshaw/wsgi-kerberos/issues/12#issuecomment-524045298).

(Of course, it also matters which clients you test with. For example, with curl you need to [explicitly add the "Transfer-Encoding: Chunked" header](https://stackoverflow.com/questions/31504531/is-it-possible-to-upload-a-file-with-curl-from-a-pipe). Also, curl is *not* the type of naive client (that blindly uploads large request bodies and expects them to be consumed before checking if authentication was successful) that we are trying to help here. See e.g. https://gms.tf/when-curl-sends-100-continue.html for more info. I guess WSGI-Kerberos does not currently support sending the 100 Continue response, but that improvement could potentially be added in a future version.)

In any case, the changes in this PR seem like strictly an improvement over the current behavior, which is to assume that a missing Content-Length header means there is a 0-length body, which is not a correct assumption.

File: wsgi_kerberos.py
Comment (propertone) [reply wsgi_kerberos.py:60]: That's fine, though I'm not sure about printing the stack trace (even in debug mode) - it might confuse developer into thinking there is an actual problem when there isn't. Should be fine to just print just the message. I'd also be explicitly about the fact that the exception comes from wsgi_kerberos so devs don't have to chase that down, eg: `wsgi_kerberos:_consume_request failed: str(exc)`

It might actually be better to just print nothing at all (as before). Unless you're a wsgi_kerberos developer, I can't imagine why devs would care about seeing debug logs of clients sending invalid content-length headers.

Comment (propertone) [reply wsgi_kerberos.py:60]: > I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here

I wouldn't rely on that - at least the wsgiref server doesn't do any validation.

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: This technically applies to 401 and 403. What do you think of `read_max_on_auth_fail`?

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: If you do leave the Exception printing, this should also cover an empty string. At least the wsgiref server converts a missing content-length header into an empty string.

**Review by propertone — CHANGES_REQUESTED** [review]

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: Done in the latest revision.

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: Thanks, handled in the latest revision.

**Comment by twosigmajab** [comment]
> > I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here
>
> I wouldn't rely on that - at least the wsgiref server doesn't do any validation.

(I think that's not surprising since wsgiref is not intended to be used in production. I meant production-quality servers, e.g. Gunicorn: https://github.com/benoitc/gunicorn/blob/548d58/gunicorn/http/message.py#L146-L150)

> I'm not sure about printing the stack trace

I originally included it because now that the exception type we're catching is sufficiently broad, when we do catch an exception, it can be helpful to see more about where it came from. That said, no longer printing the traceback in the latest revision.

> It might actually be better to just print nothing at all (as before). Unless you're a wsgi_kerberos developer, I can't imagine why devs would care about seeing debug logs of clients sending invalid content-length headers.

When swallowing exceptions like this, I think it's best practice in general to log them at debug level rather than swallow them completely silently. If there is a bug in the exception handling (e.g. wrong exception types being caught), it gives users a way of finding out about it (without having to monkey patch or modify the installed package). And users shouldn't mind when a library writes debug log messages at debug level, that's what debug level is for:)

---------

Thanks for the continued careful review! How do you think it's looking now?

File: wsgi_kerberos.py
Comment (propertone) [inline wsgi_kerberos.py]: Nit: `_consume_request suppressed error: %s`

File: test_wsgi_kerberos.py
Comment (propertone) [inline test_wsgi_kerberos.py]: This test doesn't appear to work as intended - it passes regardless of `read_max_on_auth_fail` value. Not sure what the problem is, but if it's fixable, we should have 2 tests for `len(body) > read_max_on_auth_fail` and `len(body) < read_max_on_auth_fail` to catch this.

**Review by propertone — CHANGES_REQUESTED** [review]
Almost there...

File: wsgi_kerberos.py
Comment (twosigmajab) [reply wsgi_kerberos.py]: 👍

File: test_wsgi_kerberos.py
Comment (twosigmajab) [reply test_wsgi_kerberos.py]: Yikes, good catch, thanks! Dug through the WebTest API a bit more and it looks like the best way to create a test request with a body that we can check how much has been read after submitting the request is to use `webtest.TestRequest` to explicitly create the test request with a `body` param. After submitting the request, we can then call `tell()` on the request body file "input" object to get the number of bytes read. Done in the latest revision, which also adds tests for a bunch of different read_max values that exercise all the possible cases. Another look?

**Review by propertone — APPROVED** [review]
Looks good!

**Pull request merged**
